templates/aws-stack.yml

```diff
@@ -1174,3 +1117,270 @@ Resources:
+      # We give instances one hour to respond to this hook else we continue
+      HeartbeatTimeout: 3600
+      DefaultResult: CONTINUE
+
+  ShutdownHookRule:
+    Type: AWS::Events::Rule
+    Properties:
+      Description: !Sub Run the shutdown time AWS SSM Automation for ${ShutdownHook}
+      EventPattern:
+        source:
+          - aws.autoscaling
+        detail-type:
+          - "EC2 Instance-terminate Lifecycle Action"
+        detail:
+          AutoScalingGroupName: [ !Ref AgentAutoScaleGroup ]
+          LifecycleHookName: [ !Ref ShutdownHook ]
+      Targets:
+        - Arn: !Sub arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:automation-definition/${BootHookAutomation}:$DEFAULT
+          RoleArn: !GetAtt EventBridgeRuleRole.Arn
+          Id: TargetSsmAutomation
+          InputTransformer:
+            InputPathsMap:
+              instanceid: "$.detail.EC2InstanceId"
+            InputTemplate: "{\"InstanceId\":[<instanceid>]}"
+
+  ShutdownHookAutomation:
+    Type: AWS::SSM::Document
+    Properties:
+      DocumentType: Automation
+      Content:
+        schemaVersion: "0.3"
+        assumeRole: !GetAtt AutomationRole.Arn
+        description: Stop the buildkite-agent, wait for it to exit, complete the launch lifecycle action
+        parameters:
+          InstanceId:
+            type: String
+          AutoScalingGroupName:
+            type: String
+            default: !Ref AgentAutoScaleGroup
+          LifecycleHook:
+            type: String
+            default: !Ref ShutdownHook
+        mainSteps:
+          - !If
+            - UseLinuxAgents
+            - name: RunCommand
+              action: aws:runCommand
+              inputs:
+                DocumentName: AWS-RunShellScript
+                InstanceIds:
+                  - "{{ InstanceId }}"
+                Parameters:
+                  executionTimeout: "3600"
+                  commands:
+                    - systemctl stop buildkite-agent
+            - name: RunCommand
+              action: aws:runCommand
+              inputs:
+                DocumentName: AWS-RunPowerShellScript
+                InstanceIds:
+                  - "{{ InstanceId }}"
+                Parameters:
+                  executionTimeout: "3600"
+                  commands:
+                    - nssm stop buildkite-agent
+          - name: CompleteLifecycleAction
+            action: aws:executeAwsApi
+            inputs:
+              Service: autoscaling
+              Api: CompleteLifecycleAction
+              AutoScalingGroupName: "{{ AutoScalingGroupName }}"
+              InstanceId: "{{ InstanceId }}"
+              LifecycleActionResult: "CONTINUE"
+              LifecycleHookName: "{{ LifecycleHook }}"
+
+  SpotInterruptionRule:
+    Type: AWS::Events::Rule
+    Properties:
+      Description: !Sub Run the spot interruption AWS SSM Automation for ${AgentAutoScaleGroup}
+      EventPattern:
```

Comment:
Since there’s no way to capture just instances that belong to the Auto Scaling group from the [event](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-interruptions.html#spot-instance-termination-notices), some tests to perform:

- [ ] What happens if an EC2 Spot instance outside the ASG experiences interruption, does the `autoscaling: TerminateInstanceInAutoScalingGroup` action fail and ignore the instance?